include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+        static bool is_invalid(const STextureData& _td)
+        {
+            return _td.wrap_x==EWM_INVALID||_td.wrap_y==EWM_INVALID;
+        }
+    } PACK_STRUCT;
+#include "irr/irrunpack.h"
+    static_assert(sizeof(STextureData)==sizeof(uint64_t), "STextureData is not 64bit!");
+
+protected:
+    using image_t = typename decltype(image_view_t::SCreationParams::image)::pointee;
+
+    using page_tab_offset_t = core::vector3du32_SIMD;
+    static page_tab_offset_t page_tab_offset_invalid() { return page_tab_offset_t(~0u,~0u,~0u); }
+
+    STextureData offsetToTextureData(const page_tab_offset_t& _offset, const ICPUImage* _img, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv)
+    {
+        STextureData texData;
+        texData.origsize_x = _img->getCreationParameters().extent.width;
+        texData.origsize_y = _img->getCreationParameters().extent.height;
+
+		texData.pgTab_x = _offset.x;
+		texData.pgTab_y = _offset.y;
+        texData.pgTab_layer = _offset.z;
+
+        //getCreationParameters().mipLevels doesnt necesarilly mean that there wasnt allocated space for higher non-existent mip levels
+        texData.maxMip = _img->getCreationParameters().mipLevels-1u-m_pgSzxy_log2;
+
+        auto ETC_to_int = [](ISampler::E_TEXTURE_CLAMP _etc) -> uint32_t {
+            switch (_etc)
+            {
+            case ISampler::ETC_REPEAT:
+                return STextureData::EWM_REPEAT;
+            case ISampler::ETC_CLAMP_TO_EDGE:
+            case ISampler::ETC_CLAMP_TO_BORDER:
+                return STextureData::EWM_CLAMP;
+            case ISampler::ETC_MIRROR:
+            case ISampler::ETC_MIRROR_CLAMP_TO_EDGE:
+            case ISampler::ETC_MIRROR_CLAMP_TO_BORDER:
+                return STextureData::EWM_MIRROR;
+            default:
+                return STextureData::EWM_INVALID;
+            }
+        };
+
+        texData.wrap_x = ETC_to_int(_wrapu);
+        texData.wrap_y = ETC_to_int(_wrapv);
+
+        return texData;
+    }
+
+    page_tab_offset_t alloc(const IImage* _img, const IImage::SSubresourceRange& _subres, uint32_t _pgtLayer)
+    {
+        const uint32_t pgtAddr2dMask = (1u<<(m_pgtabSzxy_log2*2u))-1u;
+
+        uint32_t szAndAlignment = computeSquareSz(_img, _subres);
+        szAndAlignment *= szAndAlignment;
+
+        uint32_t addr = pg_tab_addr_alctr_t::invalid_address;
+        core::address_allocator_traits<pg_tab_addr_alctr_t>::multi_alloc_addr((*m_pageTableLayerAllocators)[_pgtLayer], 1u, &addr, &szAndAlignment, &szAndAlignment, nullptr);
+        return (addr==pg_tab_addr_alctr_t::invalid_address) ? 
+            page_tab_offset_invalid() :
+            page_tab_offset_t(core::morton2d_decode_x(addr&pgtAddr2dMask), core::morton2d_decode_y(addr&pgtAddr2dMask), _pgtLayer);
+    }
+    virtual bool free(const STextureData& _addr, const IImage* _img, const IImage::SSubresourceRange& _subres)
+    {
+        uint32_t sz = computeSquareSz(_img, _subres);
+        sz *= sz;
```

Comment:
you can get the size from the `STextureData` (original size is saved in there) so you don't need the `const IImage* _img, const IImage::SSubresourceRange& _subres`

Reply:
still a thing

Reply:
i need to know format class of the image to find its physical storage and remove pagetable layer assignment from it. You're right i dont necesarilly need _subres though

Reply:
we should keep track of the `formatID` and `E_FORMAT` (or image view) assigned to the VT layer, would make the whole API a lot simpler